### ISSUE TITLE
Not suitable for use with http.Serve

### ISSUE THREAD

**Issue by glasser** [armon/go-proxyproto#1]
This isn't exactly a bug with this package, but we learned the hard way in production that this code (or anything implementing this interface) isn't suitable for use with `http.Server.Serve`.  Might be worth mentioning in docs.

Specifically: `http.Server.Serve` looks like:

``` go

// Serve accepts incoming connections on the Listener l, creating a
// new service goroutine for each.  The service goroutines read requests and
// then call srv.Handler to reply to them.
func (srv *Server) Serve(l net.Listener) error {
    defer l.Close()
    var tempDelay time.Duration // how long to sleep on accept failure
    for {
        rw, e := l.Accept()
        if e != nil {
            if ne, ok := e.(net.Error); ok && ne.Temporary() {
                if tempDelay == 0 {
                    tempDelay = 5 * time.Millisecond
                } else {
                    tempDelay *= 2
                }
                if max := 1 * time.Second; tempDelay > max {
                    tempDelay = max
                }
                srv.logf("http: Accept error: %v; retrying in %v", e, tempDelay)
                time.Sleep(tempDelay)
                continue
            }
            return e
        }
        tempDelay = 0
        c, err := srv.newConn(rw)
        if err != nil {
            continue
        }
        c.setState(c.rwc, StateNew) // before Serve can return
        go c.serve()
    }
}
```

`srv.newConn` calls `rw.RemoteAddr()`, and if that's a proxyproto listener, it can _block_ if the incoming connection doesn't actually send over any data.  And this is _before_ `go c.serve()`, so it's in the main `http.Server.Serve` goroutine!

Workarounds for this seem pretty complex, unfortunately...

**Comment by armon** [comment]
Hmm that is nasty. Not sure I see an obvious solution to this, it looks like it will always require ugly hacks.

**Comment by glasser** [comment]
Here's a CL for Go that should fix this. https://go-review.googlesource.com/#/c/15832/

**Comment by glasser** [comment]
and a workaround we're using: https://gist.github.com/glasser/ac531523a027cbe3955d

**Comment by glasser** [comment]
I got a simpler CL into Go: https://github.com/golang/go/commit/b58515baba3fedac7ef7ce1c3a8da600460b6c70

So this should work without a workaround in 1.6.

Issue in go repo: https://github.com/golang/go/issues/12943

**Comment by armon** [comment]
@glasser Thanks for the heads up!

**Comment by owenthereal** [comment]
@glasser I'm curious what you ended up doing with this issue? I will need to handle proxy protocol in Go too...

**Comment by glasser** [comment]
As described above, this should be fixed in Go 1.6.  My gist above is an annoying workaround for 1.5.

**Comment by owenthereal** [comment]
Cool thanks @glasser
